arroyo/dead_letter_queue/policies/count.py

```diff
@@ -0,0 +1,70 @@
+from collections import deque
+from time import time
+from typing import Callable, NamedTuple, Optional, Sequence, Tuple
+
+from arroyo.dead_letter_queue.policies.abstract import (
+    DeadLetterQueuePolicy,
+    InvalidMessage,
+)
+from arroyo.types import Message, TPayload
+from arroyo.utils.metrics import get_metrics
+
+
+class _Bucket(NamedTuple):
+    """
+    1-second time bucket to count number of hits in this second.
+    """
+
+    timestamp: int
+    hits: int
+
+
+class CountInvalidMessagePolicy(DeadLetterQueuePolicy[TPayload]):
+    """
+    Ignore invalid messages up to a certain limit per time unit window in seconds.
+    This window is 1 minute by default.
+
+    If the state of counted hits is to be persisted outside of the DLQ, a callback
+    function may be passed accepting a timestamp to add a hit to.
+
+    A saved state in the form `[(<timestamp: int>, <hits: int>), ...]` can be passed
+    on init to load the previously saved state. This state should be aggregated to
+    1 second buckets.
+    """
+
+    def __init__(
+        self,
+        limit: int,
+        seconds: int = 60,
+        load_state: Sequence[Tuple[int, int]] = [],
+        add_hit_callback: Optional[Callable[[int], None]] = None,
```

Comment:
I see. Did you consider just adding the more complex CountBackedByRedis policy in Snuba directly and letting it manage Redis state since it seems policies are designed to be extensible. Just a thought as this seems to me rather customized to Snuba's requirements already. The other thing is that we would need to get a fresh count to pass to this strategy not just when the consumer restarts but also in the case of rebalances. That part may be easier if it's all in the strategy as this can be done in the constructor of the policy class.

Reply:
I did have 2 separate classes, a `CountPolicy` and a `StatefulCountPolicy` (which extended the former and had the callback logic), but I had them both in Arroyo so I merged them together. I didn't consider putting the Redis backed count policy in Snuba but it makes sense, I'll split it back up.

Why do we need to refresh the count on a rebalance? Wouldn't we want to keep track of total number of invalid messages regardless of consumer?

Reply:
It's just that a rebalance causes the strategy to get recreated. I assumed that would recreate the policy so it needs to get a different count passed in `load_state` (the latest one)